attackcti/attack_api.py

```diff
@@ -9,27 +9,49 @@
 # https://github.com/oasis-open/cti-python-stix2/issues/183
 # https://stackoverflow.com/a/4406521
 
-from stix2 import TAXIICollectionSource, Filter, CompositeDataSource
+from stix2 import TAXIICollectionSource, Filter, CompositeDataSource, FileSystemSource
 from stix2.utils import get_type_from_id
 from taxii2client import Collection
 import json
+import os
 
 ATTCK_STIX_COLLECTIONS = "https://cti-taxii.mitre.org/stix/collections/"
 ENTERPRISE_ATTCK = "95ecc380-afe9-11e4-9b6c-751b66dd541e"
 PRE_ATTCK = "062767bd-02d2-4b72-84ba-56caef0f8658"
 MOBILE_ATTCK = "2f669986-b40b-4423-b720-4396ca6a462b"
 
+ENTERPRISE_ATTCK_LOCAL_DIR = "enterprise-attack"
+PRE_ATTCK_LOCAL_DIR = "pre-attack"
+MOBILE_ATTCK_LOCAL_DIR = "mobile-attack"
+
+
 class attack_client(object):
-    ENTERPRISE_COLLECTION = Collection(ATTCK_STIX_COLLECTIONS + ENTERPRISE_ATTCK + "/")
-    PRE_COLLECTION = Collection(ATTCK_STIX_COLLECTIONS + PRE_ATTCK + "/")
-    MOBILE_COLLECTION = Collection(ATTCK_STIX_COLLECTIONS + MOBILE_ATTCK + "/")
-    
-    TC_ENTERPRISE_SOURCE = TAXIICollectionSource(ENTERPRISE_COLLECTION)
-    TC_PRE_SOURCE = TAXIICollectionSource(PRE_COLLECTION)
-    TC_MOBILE_SOURCE = TAXIICollectionSource(MOBILE_COLLECTION)
+    TC_ENTERPRISE_SOURCE = None
+    TC_PRE_SOURCE = None
+    TC_MOBILE_SOURCE = None
+    COMPOSITE_DS = None
+
+    def __init__(self, local_path=None):
+        if local_path is not None and os.path.isdir(os.path.join(local_path, ENTERPRISE_ATTCK_LOCAL_DIR)) \
+                                  and os.path.isdir(os.path.join(local_path, PRE_ATTCK_LOCAL_DIR)) \
+                                  and os.path.isdir(os.path.join(local_path, MOBILE_ATTCK_LOCAL_DIR)):
+            self.TC_ENTERPRISE_SOURCE = FileSystemSource(os.path.join(local_path, ENTERPRISE_ATTCK_LOCAL_DIR))
```

Comment:
I don't think we need the additional `self.` on this line and LINE (39,40,49,50,51,53,54) Because it does not call methods or attributes of those libraries. All the script is doing in those steps is setting variables and initializing Collection objects and Composite Data Source objects.

Reply:
We can't leave the `self.` because that's the reference to the class (object) variables. Because I added the constructor, I need to reference to the variables with `self.` just like you do in the rest of your code. When I remove the `self.`, the variables become local variables of the constructor and the class variables remain None. Or do I misunderstand you?

Reply:
mmmm. the current code does not have the `self.` in the references to the class objects and the objects work fine right?

Reply:
That's true, in the current code you are using the variables on class level right away without a constructor. In that case you declare and initialise the variables at class level and there you don't need the `self.`. In the current code you already reference to the variables with `self.` within a function to reference to the class level variables. I added a constructor the the class to make it possible to instantiate an object from the attack_client class using the local_path parameter. And because the constructor is a function, you need to use `self.` to reference to the class level variables.

Reply:
Awesome! Thank you man for the additional information 👍